projects/baby_driver/test/test_i2c_write.c

```diff
@@ -0,0 +1,247 @@
+  reg_test = 0;
+  tx_data_test = tx_data;
+  tx_len_test = tx_len;
+  return STATUS_CODE_OK;
+}
+
+StatusCode TEST_MOCK(i2c_write_reg)(I2CPort i2c, I2CAddress addr, uint8_t reg, uint8_t *tx_data,
+                                    size_t tx_len) {
+  i2c_test = i2c;
+  addr_test = addr;
+  reg_test = reg;
+  tx_data_test = tx_data;
+  tx_len_test = tx_len;
+  return STATUS_CODE_OK;
+}
+
+static StatusCode prv_rx_i2c_write_callback(uint8_t data[8], void *context, bool *tx_result) {
+  s_times_callback_called++;
+  memcpy(s_received_data, data, 8);
+  s_received_context = context;
+  *tx_result = false;
+  return s_status_return;
+}
+
+void setup_test(void) {
+  initialize_can_and_dependencies(&s_can_storage, SYSTEM_CAN_DEVICE_BABYDRIVER, TEST_CAN_EVENT_TX,
+                                  TEST_CAN_EVENT_RX, TEST_CAN_EVENT_FAULT);
+  TEST_ASSERT_OK(dispatcher_init());
+  TEST_ASSERT_OK(i2c_write_init());
+  s_times_callback_called = 0;
+  memset(s_received_data, 0, sizeof(s_received_data));
+  s_status_return = STATUS_CODE_OK;
+
+  TEST_ASSERT_OK(
+      dispatcher_register_callback(BABYDRIVER_MESSAGE_STATUS, prv_rx_i2c_write_callback, NULL));
+}
+
+void teardown_test(void) {}
+
+// Test that data is correctly written over I2C
+void test_writing_over_i2c(void) {
+  // Test receiving multiple messages and writing over i2c with given register
+  uint8_t command_1[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, I2C_PORT_1, 0x74, 14, 2, 3, 0, 0 };
+  uint8_t data_1[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_DATA, 10, 22, 5, 4, 3, 76, 120 };
+  uint8_t data_2[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_DATA, 43, 2, 3, 4, 5, 1, 8 };
+
+  CAN_TRANSMIT_BABYDRIVER(command_1[0], command_1[1], command_1[2], command_1[3], command_1[4],
+                          command_1[5], command_1[6], command_1[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_1[0], data_1[1], data_1[2], data_1[3], data_1[4], data_1[5],
+                          data_1[6], data_1[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_2[0], data_2[1], data_2[2], data_2[3], data_2[4], data_2[5],
+                          data_2[6], data_2[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  TEST_ASSERT_EQUAL(I2C_PORT_1, i2c_test);
+  TEST_ASSERT_EQUAL(0x74, addr_test);
+  TEST_ASSERT_EQUAL(3, reg_test);
+  TEST_ASSERT_EQUAL(14, tx_len_test);
+  for (uint8_t i = 0; i < 7; i++) {
+    TEST_ASSERT_EQUAL(data_1[i + 1], tx_data_test[i]);
+  }
+  for (uint8_t j = 0; j < 7; j++) {
+    TEST_ASSERT_EQUAL(data_2[j + 1], tx_data_test[j + 7]);
+  }
```

Comment:
Pro tip: you can instead use `TEST_ASSERT_EQUAL_UINT8_ARRAY(expected, actual num_elements)`, like `TEST_ASSERT_EQUAL_UINT8_ARRAY(data_1 + 1, tx_data_test, 7)`.